RFC Throw exception or print error to stdout (#9)
Vindaar/shell: README.org

** Debugging
In order to see what's going on, you can either compile your program
with the =-d:debugShell= flag, which will then echo the rewritten
commands during compilation.
Alternatively in order to avoid calling the commands immediately, you
may use the =shellEcho= macro instead. It simply echoes the commands
that would otherwise be run.

** Error reporting

By default ~shell~ prints output messages to stdout:

#+BEGIN_SRC nim :exports both :results scalar
import shell

shell:
  ls
#+END_SRC

#+RESULTS:
: shellCmd: ls
: shell> nim.cfg
: shell> README.org
: shell> shell
: shell> shell.nim
: shell> shell.nim.bin
: shell> shell.nimble
: shell> tests

What is printed to stdout can be configured by using defines:

- ~shellNoDebugOutput~ :: Do not print command output
- ~shellNoDebugError~ :: Do not print error output
- ~shellNoDebugCommand~ :: Do not print command being executed
- ~shellNoDebugRuntime~ :: When error occurs do not print failed command

> By default the are disabled - to enable use either ~-d:shellNoDebug*~

Reviewer: ```suggestion
By default these are disabled - to enable use either ~-d:shellNoDebug*~
```

> or use ~{.define(shellNoDebug*).}~ pragma

Reviewer: ```suggestion
or use the ~{.define(shellNoDebug*).}~ pragma.
```

#+BEGIN_SRC nim :exports both :results scalar
{.define(shellNoDebugOutput).}

import shell

shell:
  ls
#+END_SRC

#+RESULTS:
: shellCmd: ls

> Default ~shellVerbose~ command combines stderr and stdout into single

Reviewer: ```suggestion
The default ~shellVerbose~ command combines stderr and stdout into single
```

> result. Tot get =stdout=, =stderr= and return code separately use

Reviewer: ```suggestion
result. To get =stdout=, =stderr= and the return code separately use
```

~shellVerboseErr~. Both of these templates have an overload that takes
~set[DebugOutputKind]~ to control printing settings:

#+BEGIN_SRC nim :exports both :results scalar
import shell

let (res, err, code) = shellVerboseErr {dokCommand}:
  echo "test"

echo "Returned string: '", res, "' with exit code ", code

#+END_SRC

#+RESULTS:
: shellCmd: echo test
: Returned string: 'test' with exit code 0

> Printing error directly into stdout is good solution for most of the

Reviewer: ```suggestion
Printing errors directly into stdout is good solution for most of the
```

use cases, but sometimes it is necessary to provide more sophisticated

> error handing - throwing exception when command failed. To switch to

Reviewer: ```suggestion
error handing - throwing an exception when the command failed. To switch to
```

exceptions use ~-d:shellThrowException~. It will automatically disable
all other output types in the default configuration.

#+begin_src nim :exports both :results scalar
{.define(shellThrowException).}

import shell, strutils

try:
  shell:
    ls -l
    ls -z
except ShellExecError:
  let e = cast[ShellExecError](getCurrentException())
  echo e.msg # Error message describing what happened
  echo "command was: ", e.cmd # Original command string
  assert e.cmd == "ls -z"
  echo "return code: ", e.retcode # Return code
  echo "error outpt: "
  for l in e.errstr.split('\n'): # Stderr from the command
    echo "  ", l
#+end_src

#+RESULTS:
: Command ls -z exited with non-zero code
: command was: ls -z
: return code: 2
: error outpt:
:   ls: invalid option -- 'z'
:   Try 'ls --help' for more information.

> On command failure ~ShellExecError~

Reviewer: ```suggestion
On command failure ~ShellExecError~ is raised.
```
